classes/core_userkey_manager.php

```diff
@@ -124,8 +124,22 @@ public function validate_key($keyvalue) {
 
         if ($key->iprestriction) {
             $remoteaddr = getremoteaddr(null);
-            if (empty($remoteaddr) or !address_in_subnet($remoteaddr, $key->iprestriction)) {
-                print_error('ipmismatch');
+            $whitelist = get_config('auth_userkey', 'ipwhitelist');
+            if (!empty($whitelist)) {
+                $ips = explode(';', $whitelist);
+                $whitelisted = false;
+                foreach ($ips as $ip) {
+                    if (address_in_subnet($remoteaddr, $ip)) {
+                        $whitelisted = true;
+                    }
+                }
+                if (!$whitelisted) {
+                    print_error('ipmismatch', 'error', '', null, "Remote address: $remoteaddr\nKey IP: $key->iprestriction");
+                }
+            } else if (empty($remoteaddr) ) {
```

Comment:
Should we check this first and throw an error if empty?

Reply:
Probably... Have changed to check empty($remoteaddr) first